consensus/src/terminator.rs

```diff
@@ -0,0 +1,121 @@
+use futures::channel::oneshot;
+use log::debug;
+
+pub type TerminatorConnection = (oneshot::Sender<()>, oneshot::Receiver<()>);
+pub type ShutdownConnection = (oneshot::Receiver<()>, Option<TerminatorConnection>);
```

Comment:
A suggestion for a possible improvement, I won't insist, but I think it would make a lot of things look much cleaner.

I'm pretty sure you could add one more `struct` (say `ShutdownManager` for the rest of this comment, but feel free to come up with a cooler/more appropriate name) and work only on these two structs.

`ShutdownManager` would be passed to every component. It would have a method `async pub fn exit(&mut self) -> &mut Terminator` which would wait for an exit signal and return the terminator when it receives one – then the terminator could be used to `terminate_sync` just as now. It would also have a method `add_offspring(&mut self, component_name: &'static str) -> ShutdownManager` which would let you immediately pass the appropriate `ShutdownManager` to a subtask (internally it would obviously call the appropriate method on its `Terminator`). Crucially `Terminator` would now contain the send exit channels as well and `terminate_sync` would start by sending exit messages to all offspring.

This would eliminate the typedefs above which result in slightly mysterious deconstruction into pairs (someone complained when I did something like this recently, and I think this is a valid complaint) and explicit passing of exit channels. It would also make a lot of the exit code in various components cleaner.

Reply:
Oh, I haven't noticed your comment that suggests something in this vein anyways, nice. ^^"

Reply:
Yes, I thought about hiding exits in Terminator, but haven't found a way to make passing channels to components look clean - this wrapper seems like a good idea. Will think it through and implement soon.